feature/haystack integration

ssean1 opened the pull request:
- Remove Elasticmodels from the project
- Add Haystack package, settings, and files to the project
- Refactor Report, Species, and Users apps to use Haystack search backend
- Refactor templates to hit the database less
- Bug fixes from changes
- Refactor tests

ssean1 commented:
I will append commits as I finish work on this project. Currently, I am in the process of refactoring tests

ssean1 commented:
Rebasing and cleaning up these commits.

ssean1 commented:
I've cleaned up this PR and added some tests, I believe this is ready for review. Most of the big changes are in 0605322's ReportSearchForm.

wylee commented:
To avoid hitting the database when doing searches, I'd recommend that you create a `ReportMixin` class that has common properties & methods for both `Report` and search results. This might require you to add some non-indexed fields to `ReportIndex` to make it look more similar to `Report`.

ssean1 commented:
@wylee Would ReportSearchResult be defined in `hotline/reports/search_indexes.py`? Same for ReportMixin?

wylee commented:
Good question. Off the top of my head, I'd say `ReportMixin` should go in the model and `ReportSearchResult` should go in `search_indexes`.

wylee commented:
Another option instead of a mixin and custom `ReportSearchResult` would be to create a `ReportSerializer` class that does the stuff that's in `Report.to_json()`. You could then call `ReportSerializer().serialize()` with either a `Report` or a `SearchResult`. This will work if you add the necessary fields to `ReportIndex`.

I actually like this approach better. It's basically how Django REST Framework does things.

ssean1 commented:
Wouldn't that require rest_framework as a new dependency? Or from what I understand use `ReportSerializer().serialize()` to do the work of what `Report.to_json()` currently does?

wylee commented:
You can create your own simple serializer w/o needing DRF. On the other hand, you _could_ add DRF as dependency and use its serializers.

ssean1 commented:
I'll add DRF as a dependency and use it, that way if we need it in the future or need to extend or change functionality, switch to class based views, use more serialization, etc, it'll already be a part of the site.

wylee commented:
+1 to using DRF

ssean1 commented:
I'm assuming the convert script needs to be refactored as well, since it uses elasticmodels?

ssean1 commented:
I removed the unused Elasticsearch connections from the settings, committed it, and rebased it into c2bdb50. I've also added the serializer here as well and integrated it into the views/templates. Going to refactor the conversion script now, feel free to look it over when you get a chance.

wylee commented:
```
diff --git a/hotline/reports/forms.py b/hotline/reports/forms.py
index c00702f..da6ee1c 100644
--- a/hotline/reports/forms.py
+++ b/hotline/reports/forms.py
@@ -7,8 +7,9 @@ from django.core.urlresolvers import reverse
 from django.core.validators import validate_email
 from django.template.loader import render_to_string
 from django.utils.safestring import mark_safe
+
 from haystack.forms import SearchForm
-from haystack.query import SQ
+from haystack.query import SQ, SearchQuerySet
 
 from hotline.comments.models import Comment
 from hotline.counties.models import County
@@ -16,7 +17,6 @@ from hotline.notifications.models import UserNotificationQuery
 from hotline.species.models import Category, Severity, Species
 from hotline.users.models import User
 
-from .search_indexes import ReportIndex
 from .models import Invite, Report
 
 
@@ -67,10 +67,21 @@ class ReportSearchForm(SearchForm):
     def __init__(self, *args, user, report_ids=(), **kwargs):
         self.user = user
         self.report_ids = report_ids
-        super().__init__(*args, **kwargs)
+
+        sqs = SearchQuerySet().models(Report)
+        if not user.is_active:
+            if self.report_ids:
+                sqs = sqs.filter(SQ(id__in=self.report_ids) | SQ(is_public=True))
+            else:
+                sqs = sqs.filter(is_public=True)
+
+        super().__init__(*args, searchqueryset=sqs, **kwargs)
+
+        self.searchqueryset = self.searchqueryset.models(Report)
 
         # only certain fields on this form can be used by members of the public
         public_fields = ['q', 'sort_by', 'source']
+
         if not user.is_active:
             for name in self.fields.keys():
                 if name not in public_fields:
@@ -108,23 +119,19 @@ class ReportSearchForm(SearchForm):
             yield category, self['category-%d' % category.pk]
 
     def no_query_found(self):
-        """
-        If no results came up from a query, or the query is invalid, return all
-        of the reports
-        """
-        return self.searchqueryset.all().models(Report)
+        """Return all reports when query is invalid or there's no search term."""
+        return self.searchqueryset.all()
 
     def search(self):
-        """
-        Searches for reports in the ES index, filters, sorts, and return results
-        as a Haystack SearchQuerySet
-        """
-        results = super().search().models(Report)
+        search_term = self.cleaned_data.get('q')
 
-        # We want to display all results when the query is not valid
-        if not self.is_valid():
+        if not self.is_valid() or not search_term:
             return self.no_query_found()
 
+        # XXX: Not sure we want to do an auto query here; e.g., we might
+        #      to boost the title field.
+        results = self.searchqueryset.auto_query(search_term)
+
         # only show public reports if the user is inactive
         if not self.user.is_active:
             if self.report_ids:
diff --git a/hotline/reports/serializers.py b/hotline/reports/serializers.py
index 365572b..5cf100a 100644
--- a/hotline/reports/serializers.py
+++ b/hotline/reports/serializers.py
@@ -12,12 +12,16 @@ class ReportSerializer(serializers.Serializer):
     so it can be passed to the template and used in our Google Maps JS logic.
     """
 
+    pk = serializers.IntegerField()
     content = serializers.SerializerMethodField()
     icon = serializers.CharField(source="icon_url")
+    icon_url = serializers.CharField(required=False)
     image_url = serializers.CharField(required=False)
     lat = serializers.SerializerMethodField()
     lng = serializers.SerializerMethodField()
     title = serializers.CharField()
+    species = serializers.CharField()
+    category = serializers.CharField()
 
     def get_content(self, instance):
         """
diff --git a/hotline/reports/views.py b/hotline/reports/views.py
index 2f9206f..228babd 100644
--- a/hotline/reports/views.py
+++ b/hotline/reports/views.py
@@ -54,14 +54,14 @@ def list_(request):
     # 1. Get the reports from the form's Search method
     # 2. Instantiate a serializer with the reports returned from form.search()
     # 3. Render the serialized data to a valid JSON object
-    reports = form.search()
-    serializer = ReportSerializer(reports, many=True)
-    reports_json = JSONRenderer().render(serializer.data)
+    results = form.search()
+    serializer = ReportSerializer(results, many=True)
+    reports = serializer.data
 
     return render(request, template, dict({
         "reports": reports,
         "form": form,
-        "reports_json": reports_json,
+        "reports_json": json.dumps(reports),
         "tab": tab,
     }, **tab_context))
```

wylee commented on hotline/reports/forms.py:
```diff
@@ -119,90 +107,83 @@ def iter_categories(self):
-                Q(pk__in=self.report_ids)
-            )
+    def search(self):
+        """
```
This docstring isn't necessary because it doesn't provide any information beyond what we already know about the `search` method.

wylee commented on hotline/reports/forms.py:
```diff
@@ -119,90 +107,83 @@ def iter_categories(self):
-            "actual_species__category",
-            "reported_species__category"
-        )
+    def no_query_found(self):
```
This should check if the user isn't active and filter out non-public reports, right? (see patch)

wylee commented:
I'm wondering if you know why on the beta site the `/reports/list` page shows a lot less reports than on the Haystack version? Does it do pagination or something?

ssean1 commented:
The old version does indeed paginate the results. I've got pagination working just now, I'll add a commit for that along with a commit or two that addresses the comments.

ssean1 commented:
Exporting the results to CSV/KML appears to be broken. I will address this

ssean1 commented:
Two issues with applying the diff:
1. In `ReportSearchForm.search()`, when `search_term` isn't present, it will return `no_query_found` which displays all results. This prevents any further filtering of the results as long as a search query is not entered. (For example, if someone wanted to narrow down the results with filters, this logic would *require* a search term to exist)
2. After applying the diff and also adding some basic pagination logic, using `reports` as serializer.data for both the JSON and template content results in the serializer failing to parse the paginator information as JSON. Would it still be reasonable to use `results` wrapped up in a paginator as what we send to the template?

wylee commented:
That patch was rough, so it's more than likely that I introduced something that's not quite right.

As far as 2, I'm not sure what you mean. You can't pass a page to the serializer if that's what you mean.

ssean1 commented:
I've added 4 commits that are new, and added some changes to 2978ec1 and f6fd675 (I just changed the search placeholder to `Enter a keyword and click "Search"`
